### PR TITLE
feat: impute the molecules an unbalanced reaction is missing

### PR DESCRIPTION
**Pull request by tagirshin**
Efficient version of SynRBL

<h3>Greptile Summary</h3>

This PR replaces CGR-only reaction rebalancing with missing-molecule imputation and integrates the new behavior into standardization. It also removes superseded filters, improves ORD desired-product selection, and adds focused tests and benchmark tooling.
- Adds formula-, fragment-, reagent-, and redox-aware reaction rebalancing with confidence-based rejection.
- Exposes the rebalancing controls through standardization configuration and updates the documented pipeline behavior.
- Selects desired ORD products independently within each outcome, preserving unflagged outcomes and preventing fallback to an unflagged sibling when a desired product lacks SMILES.
- Removes the competing-product and disconnected-CGR filters superseded by the new workflow.

<h3>Confidence Score: 5/5</h3>

The PR appears safe to merge.

No blocking failure remains.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| synplan/chem/data/rebalancing.py | Implements the missing-molecule imputation engine, confidence calculation, redox handling, fragment recovery, and balance validation. |
| synplan/chem/data/standardizing.py | Integrates rebalancing into the standardization configuration and execution order after reagent removal. |
| synplan/utils/ord/reader.py | Resolves desired products within each outcome while retaining all products from outcomes without a desired-product flag. |
| synplan/chem/data/filtering.py | Removes the competing-product and CGR-connected-components filters superseded by the revised curation flow. |
| tests/unit/chem/data/test_rebalancing.py | Exercises the new balancing strategies, configuration controls, and chemical edge cases. |
| tests/unit/utils/test_ord_reader.py | Covers desired-product selection, unflagged outcomes, and desired products without structural identifiers. |
| docs/configuration/standardization.rst | Documents all supported rebalancing options, including unsupported-redox refusal. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart LR
    A[ORD or reaction input] --> B[Parse reaction record]
    B --> C[Select products per ORD outcome]
    C --> D[Remove reagents]
    D --> E{Reaction balanced?}
    E -- Yes --> F[Continue standardization]
    E -- No --> G[Impute missing molecules]
    G --> H{Confidence and redox policy accepted?}
    H -- Yes --> F
    H -- No --> I[Exclude reaction]
    F --> J[Filtering and curated output]
```
</details>

<sub>Reviews (4): Last reviewed commit: ["test: assert the composed CGR rather tha..."](https://github.com/laboratoire-de-chemoinformatique/synplanner/commit/930d431a31b8ce1fe258f4931fd8caacb1c71815) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=51233775)</sub>

<details><summary><h4>Context used (3)</h4></summary>

- Knowledge Base — [Chemical reaction data pipeline](https://app.greptile.com/tagirshin/-/custom-context/knowledge-base/laboratoire-de-chemoinformatique/synplanner/-/docs/chemical-data-pipeline.md)
- Knowledge Base — [Interfaces, configuration, and utilities](https://app.greptile.com/tagirshin/-/custom-context/knowledge-base/laboratoire-de-chemoinformatique/synplanner/-/docs/interfaces-configuration-and-utilities.md)
- Knowledge Base — [Reaction data curation](https://app.greptile.com/tagirshin/-/custom-context/knowledge-base/laboratoire-de-chemoinformatique/synplanner/-/docs/reaction-data-curation.md)
</details>

### CODE REVIEW

File: synplan/chem/utils.py [hunk 1]
```diff
@@ -165,6 +166,26 @@ def is_reaction_atom_mapped(reaction: ReactionContainer | str) -> bool:
     return reaction_mapping_status(reaction) == "fully_mapped"
 
 
+def strip_reaction_mapping(reaction: ReactionContainer) -> ReactionContainer:
```
Comment (Protolaw) [inline synplan/chem/utils.py]: > I’m not sure we need this extra helper. `chython-synplan` already covers most of the basic AAM operations through `ReactionContainer`:
>
> ```python
> str(reaction)                     # map-free SMILES text
> format(reaction, "m")             # SMILES including atom maps
> smiles(str(reaction))             # new, genuinely unmapped container
> smiles(text, remap=True)          # renumber while preserving mapping
> reaction.reset_mapping()          # compute/reset the mapping
> ```
>
> I understand that the helper avoids the serialization/reparse cost of `smiles(str(reaction))`, but currently it seems to be used only in tests and the rebalancing benchmark. For rebalancing, we already have `use_mapping=False`, which expresses the intention more directly without modifying the input reaction.
>
> The in-place mutation also worries me. `~reaction` is cached by Chython, but remapping the individual molecules does not clear the reaction-level cache. For example:
>
> ```python
> cgr_before = ~reaction
> strip_reaction_mapping(reaction)
> cgr_after = ~reaction
>
> assert cgr_after is cgr_before  # stale mapped CGR
> ```
>
> We would at least need:
>
> ```python
> reaction.flush_cache(keep_molecule_cache=True)
> ```
>
> IMO it can be even worse than just returning stale data: the helper destructively removes the cross-side atom correspondence from the original object, so we lose the information needed to identify the reaction center. If the caller only wants to ignore mapping for one operation, `use_mapping=False` or creating a separate unmapped container feels safer than changing the original reaction globally.

Comment (tagirshin) [reply synplan/chem/utils.py]: Agreed, I will remove it

File: synplan/chem/data/rebalancing.py [hunk 2]
```diff
@@ -0,0 +1,1375 @@
...
+
+
+def _small_molecules(smiles_strings: list[str]) -> list[MoleculeContainer]:
+    """Parse the species to add.
+
+    ``[O]`` is read with ``keep_implicit`` because chython otherwise fills the
+    valence of a bare bracket atom and hands back water.
+    """
+    return [
+        smiles_chython(text, keep_implicit=text == ATOMIC_OXYGEN)
```
Comment (Protolaw) [inline synplan/chem/data/rebalancing.py]: Blocking: could we make this survive the actual SMI writer/reader round trip before merging?

`keep_implicit=True` is container state that is not preserved by `format(reaction, "m")`. For both `CC>>C=C` and `CC(O)C>>CC(=O)C`, the result is balanced in memory, but the written `[O:n]` is parsed back as `[OH2:n]`, leaving `{"H": 2}` imbalance.

I think we either need a representation that round-trips as atomic oxygen, or should avoid emitting atomic oxygen altogether. A write/read balance regression test would catch this at the production boundary.

File: synplan/chem/data/rebalancing.py [hunk 3]
```diff
@@ -0,0 +1,1375 @@
...
+    reactant numbers that say where it came from.
+
+    Reagents count as part of the reactant side: that is where
+    ``ReactionContainer.compose`` puts them.
+    """
+    spare = max((n for m in reaction.molecules() for n in m), default=0)
+
+    def renumber(molecule: MoleculeContainer, clash: set[int]) -> None:
+        nonlocal spare
+        molecule.remap({n: spare + i for i, n in enumerate(clash, 1)})
```
Comment (Protolaw) [inline synplan/chem/data/rebalancing.py:1223]: Blocking for CGR/rule extraction: collision-free numbering is not the same as valid AAM.

In mapped ethyl-benzoate hydrolysis, the added reactant water oxygen gets map 12 while the product ethanol oxygen gets map 13. Methyl-acetate hydrolysis similarly produces O6/O7. Both outputs become partially mapped, `fix_mapping()` returns false, and the CGR represents a conserved oxygen as disappearance plus appearance.

Could we establish correspondence for atoms transferred between the imputed species, or run and validate a full remapping before returning? Merely making `~reaction` composable gives the wrong reaction-center identity.

File: synplan/chem/data/rebalancing.py (hunk 2 above)
Comment (tagirshin) [inline synplan/chem/data/rebalancing.py]: Hm, it is a part of standardization workflow, so it makes sense to add smiles roundtrip as a standardisation step. Or you mean add specifically here as default?

File: synplan/chem/data/rebalancing.py (hunk 3 above)
Comment (tagirshin) [reply synplan/chem/data/rebalancing.py:1223]: Wait, it is made for non mapped reaction. The SynRBL works with no mapping at all, so here we omit it, if ignore mapping = true. But if not, then this function should not activate. Or what are you asking here?

**Comment by tagirshin** [comment]
@greptileai review

**Comment by tagirshin** [comment]
@greptileai review

**Comment by tagirshin** [comment]
Superseded by #94, which contains every commit from this branch — `feature/rebalancing` is a complete ancestor of `feature/synton`, so nothing here is lost.

Rebalancing and the synthon toolkit ship together as 1.7.0 rather than as two releases, so this branch does not need its own merge. Closing to free #94 to retarget onto `main`.
